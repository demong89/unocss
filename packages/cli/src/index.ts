import { existsSync, promises as fs } from 'fs'
import { basename, dirname, relative, resolve } from 'pathe'
import fg from 'fast-glob'
import consola from 'consola'
import { cyan, dim, green } from 'colorette'
import { debounce } from 'perfect-debounce'
import { toArray } from '@unocss/core'
import type { SourceCodeTransformerEnforce, UserConfig } from '@unocss/core'
import { version } from '../package.json'
import { createContext } from '../../shared-integration/src/context'
import { applyTransformers } from '../../shared-integration/src/transformers'
import { PrettyError, handleError } from './errors'
import { defaultConfig } from './config'
import type { CliOptions, ResolvedCliOptions } from './types'

const name = 'unocss'

export async function resolveOptions(options: CliOptions) {
  if (!options.patterns?.length) {
    throw new PrettyError(
      `No glob patterns, try ${cyan(`${name} <path/to/**/*>`)}`,
    )
  }

  return options as ResolvedCliOptions
}

export async function build(_options: CliOptions) {
  const fileCache = new Map<string, string>()

  const cwd = _options.cwd || process.cwd()
  const options = await resolveOptions(_options)

  async function loadConfig() {
    const ctx = createContext<UserConfig>(options.config, defaultConfig)
    const configSources = (await ctx.updateRoot(cwd)).sources
    return { ctx, configSources }
  }

  const { ctx, configSources } = await loadConfig()
  const files = await fg(options.patterns, { cwd, absolute: true })
  await Promise.all(
    files.map(async (file) => {
      fileCache.set(file, await fs.readFile(file, 'utf8'))
    }),
  )

  consola.log(green(`${name} v${version}`))
  consola.start(`UnoCSS ${options.watch ? 'in watch mode...' : 'for production...'}`)

  const debouncedBuild = debounce(
    async () => {
      generate(options).catch(handleError)
    },
    100,
  )

  const startWatcher = async () => {
    if (!options.watch)
      return

    const { watch } = await import('chokidar')
    const { patterns } = options
    const ignored = ['**/{.git,node_modules}/**']

    const watcher = watch(patterns, {
      ignoreInitial: true,
      ignorePermissionErrors: true,
      ignored,
      cwd,
    })

    if (configSources.length)
      watcher.add(configSources)

    watcher.on('all', async (type, file) => {
      const absolutePath = resolve(cwd, file)

      if (configSources.includes(absolutePath)) {
        await ctx.reloadConfig()
        consola.info(`${cyan(basename(file))} changed, setting new config`)
      }
      else {
        consola.log(`${green(type)} ${dim(file)}`)

        if (type.startsWith('unlink'))
          fileCache.delete(absolutePath)
        else
          fileCache.set(absolutePath, await fs.readFile(absolutePath, 'utf8'))
      }

      debouncedBuild()
    })

    consola.info(
      `Watching for changes in ${
        toArray(patterns)
          .map(i => cyan(i))
          .join(', ')}`,
    )
  }

  await generate(options)

  await startWatcher().catch(handleError)

  function transformFiles(sources: { id: string; code: string; transformedCode?: string | undefined }[], enforce: SourceCodeTransformerEnforce = 'default') {
    return Promise.all(
      sources.map(({ id, code, transformedCode }) => new Promise<{ id: string; code: string; transformedCode: string | undefined }>((resolve) => {
        applyTransformers(ctx, code, id, enforce)
          .then((transformsRes) => {
            resolve({ id, code, transformedCode: transformsRes?.code || transformedCode })
          })
      })))
  }

  async function generate(options: ResolvedCliOptions) {
    const sourceCache = Array.from(fileCache).map(([id, code]) => ({ id, code }))

    const outFile = resolve(options.cwd || process.cwd(), options.outFile ?? 'uno.css')

    const preTransform = await transformFiles(sourceCache, 'pre')
    const defaultTransform = await transformFiles(preTransform)
    const postTransform = await transformFiles(defaultTransform, 'post')

    // update source file
    await Promise.all(
      postTransform.filter(({ transformedCode }) => !!transformedCode)
        .map(({ transformedCode, id }) => new Promise<void>((resolve) => {
          if (existsSync(id))
            fs.writeFile(id, transformedCode as string, 'utf-8').then(resolve)
        })),
    )

    const { css, matched } = await ctx.uno.generate(
      [...postTransform.map(({ code, transformedCode }) => transformedCode ?? code)].join('\n'),
      {
        preflights: options.preflights,
        minify: options.minify,
      },
    )

    const dir = dirname(outFile)
    if (!existsSync(dir))
      await fs.mkdir(dir, { recursive: true })
    await fs.writeFile(outFile, css, 'utf-8')

    if (!options.watch) {
      consola.success(
      `${[...matched].length} utilities generated to ${cyan(
        relative(process.cwd(), outFile),
      )}\n`,
      )
    }
  }
}
