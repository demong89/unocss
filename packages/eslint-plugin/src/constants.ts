export const CLASS_FIELDS = ['class', 'classname']
