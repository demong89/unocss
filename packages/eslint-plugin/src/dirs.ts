import { fileURLToPath } from 'url'

export const distDir = fileURLToPath(new URL('../dist', import.meta.url))
