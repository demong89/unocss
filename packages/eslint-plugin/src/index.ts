import order from './rules/order'
import orderAttributify from './rules/order-attributify'
import configsRecommended from './configs/recommended'

export default {
  rules: {
    order,
    'order-attributify': orderAttributify,
  },
  configs: {
    recommended: configsRecommended,
  },
}
