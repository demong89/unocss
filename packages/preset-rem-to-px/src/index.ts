import type { Preset } from '@unocss/core'

const remRE = /(-?[\.\d]+)rem/g

export interface RemToPxOptions {
  /**
   * 1rem = n px
   * @default 16
   */
  baseFontSize?: number
}

export default function remToPxPreset(options: RemToPxOptions = {}): Preset {
  const {
    baseFontSize = 16,
  } = options

  return {
    name: '@unocss/preset-rem-to-px',
    postprocess: (util) => {
      util.entries.forEach((i) => {
        const value = i[1]
        if (value && typeof value === 'string' && remRE.test(value))
          i[1] = value.replace(remRE, (_, p1) => `${p1 * baseFontSize}px`)
      })
    },
  }
}
