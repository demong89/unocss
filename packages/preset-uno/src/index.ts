import type { Preset } from '@unocss/core'
import type { PresetMiniOptions, Theme } from '@unocss/preset-mini'
import { rules, shortcuts, theme, variants } from '@unocss/preset-wind'
import { VarPrefixPostprocessor, normalizePreflights, preflights } from '@unocss/preset-mini'
import { variantColorMix } from './variants/mix'

export type { Theme }

export interface PresetUnoOptions extends PresetMiniOptions {}

export const presetUno = (options: PresetUnoOptions = {}): Preset<Theme> => {
  options.dark = options.dark ?? 'class'
  options.attributifyPseudo = options.attributifyPseudo ?? false
  options.preflight = options.preflight ?? true
  options.variablePrefix = options.variablePrefix ?? 'un-'

  return {
    name: '@unocss/preset-uno',
    theme,
    rules,
    shortcuts,
    variants: [
      ...variants(options),
      variantColorMix(),
    ],
    options,
    postprocess: VarPrefixPostprocessor(options.variablePrefix),
    preflights: options.preflight ? normalizePreflights(preflights, options.variablePrefix) : [],
    prefix: options.prefix,
  }
}

export default presetUno
