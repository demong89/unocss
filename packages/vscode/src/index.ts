import path from 'path'
import type { ExtensionContext, StatusBarItem } from 'vscode'
import { StatusBarAlignment, commands, window, workspace } from 'vscode'
import { version } from '../package.json'
import { log } from './log'
import { registerAnnotations } from './annotation'
import { registerAutoComplete } from './autocomplete'
import { ContextLoader } from './contextLoader'
import { registerSelectionStyle } from './selectionStyle'

const registerRoot = async (ext: ExtensionContext, status: StatusBarItem, cwd: string) => {
  const contextLoader = new ContextLoader(cwd)

  await contextLoader.ready

  const hasConfig = await contextLoader.loadContextInDirectory(cwd)
  // TODO: improve this to re-enable after configuration created
  if (!hasConfig)
    throw new Error(`‚ûñ No config found in ${cwd}`)

  registerAutoComplete(cwd, contextLoader, ext)
  registerAnnotations(cwd, contextLoader, status, ext)
  registerSelectionStyle(cwd, contextLoader)

  return contextLoader
}

export async function activate(ext: ExtensionContext) {
  log.appendLine(`‚ö™Ô∏è UnoCSS for VS Code v${version}\n`)

  const projectPath = workspace.workspaceFolders?.[0].uri.fsPath
  if (!projectPath) {
    log.appendLine('‚ûñ No active workspace found, UnoCSS is disabled')
    return
  }

  const config = workspace.getConfiguration('unocss')
  const disabled = config.get<boolean>('disable', false)
  if (disabled) {
    log.appendLine('‚ûñ Disabled by configuration')
    return
  }

  const status = window.createStatusBarItem(StatusBarAlignment.Right, 200)
  status.text = 'UnoCSS'

  const root = config.get<string | string[]>('root')

  if (Array.isArray(root) && root.length) {
    const cwds = root.map(dir => path.resolve(projectPath, dir))

    try {
      const contextLoaders = await Promise.all(cwds.map(cwd => registerRoot(ext, status, cwd)))

      ext.subscriptions.push(
        commands.registerCommand('unocss.reload', async () => {
          log.appendLine('üîÅ Reloading...')
          await Promise.all(contextLoaders.map(ctxLoader => ctxLoader.reload))
          log.appendLine('‚úÖ Reloaded.')
        }),
      )
    }
    catch (error) {
      log.appendLine((error as Error).message)
    }
    return
  }

  // now if the root is an array, then it is an empty array
  const cwd = root && !Array.isArray(root)
    ? path.resolve(projectPath, root)
    : projectPath

  try {
    const contextLoader = await registerRoot(ext, status, cwd)

    ext.subscriptions.push(
      commands.registerCommand('unocss.reload', async () => {
        log.appendLine('üîÅ Reloading...')
        await contextLoader.reload()
        log.appendLine('‚úÖ Reloaded.')
      }),
    )
  }
  catch (error) {
    log.appendLine((error as Error).message)
  }
}

export function deactivate() {}
