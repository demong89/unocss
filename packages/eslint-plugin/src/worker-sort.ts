import { loadConfig } from '@unocss/config'
import type { UnoGenerator } from '@unocss/core'
import { createGenerator } from '@unocss/core'
import { runAsWorker } from 'synckit'
import { sortRules } from '../../shared-integration/src/sort-rules'

async function getGenerator() {
  const { config } = await loadConfig()
  return createGenerator(config)
}

let promise: Promise<UnoGenerator<any>> | undefined

runAsWorker(async (classes: string) => {
  promise = promise || getGenerator()
  const uno = await promise
  return await sortRules(classes, uno)
})
