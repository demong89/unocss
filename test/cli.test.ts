import { resolve } from 'path'
import fs from 'fs-extra'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import { startCli } from '../packages/cli/src/cli-start'

export const tempDir = resolve('.temp')
export const cli = resolve(__dirname, '../packages/cli/src/cli.ts')

beforeAll(async () => {
  await fs.remove(tempDir)
})

afterAll(async () => {
  await fs.remove(tempDir)
})

describe('cli', () => {
  it('builds uno.css', async () => {
    const { output } = await runCli({
      'views/index.html': '<div class="p-4 max-w-screen-md"></div>',
    })

    expect(output).toMatchSnapshot()
  })

  it('supports unocss.config.js', async () => {
    const { output } = await runCli({
      'views/index.html': '<div class="box"></div>',
      'unocss.config.js': `
import { defineConfig } from 'unocss'
export default defineConfig({
  shortcuts: [{ box: 'max-w-7xl mx-auto bg-gray-100 rounded-md shadow-sm p-4' }]
})
    `.trim(),
    })

    expect(output).toMatchSnapshot()
  })

  it('supports variantGroup transformer', async () => {
    const { output, transform } = await runCli({
      'views/index.html': '<div class="p-4 border-(solid red)"></div>',
      'unocss.config.js': `
import { defineConfig, transformerVariantGroup } from 'unocss'
export default defineConfig({
  transformers: [transformerVariantGroup()]
})
    `.trim(),
    }, 'views/index.html')
    expect(output).toMatchSnapshot()
    expect(transform).toMatchSnapshot()
  })

  it('supports directives transformer', async () => {
    const { output, transform } = await runCli({
      'views/index.css': '.btn-center{@apply text-center my-0 font-medium;}',
      'unocss.config.js': `
import { defineConfig, transformerDirectives } from 'unocss'
export default defineConfig({
  transformers: [transformerDirectives()]
})
    `.trim(),
    }, 'views/index.css')
    expect(output).toMatchSnapshot()
    expect(transform).toMatchSnapshot()
  })

  it('uno.css exclude initialized class after changing file', async () => {
    const fileName = 'views/index.html'
    const initializedContent = '<div class="bg-blue"></div>'
    const changedContent = '<div class="bg-red"></div>'
    const testDir = getTestDir()
    const absolutePathOfFile = resolve(testDir, fileName)
    await fs.outputFile(absolutePathOfFile, initializedContent)
    await runAsyncChildProcess(testDir, './views/**/*', '-w')
    const outputPath = resolve(testDir, 'uno.css')
    for (let i = 50; i >= 0; i--) {
      await sleep(50)
      if (fs.existsSync(outputPath))
        break
    }
    await fs.writeFile(absolutePathOfFile, changedContent)
    // polling until update
    for (let i = 100; i >= 0; i--) {
      await sleep(100)
      const output = await readFile(testDir)
      if (i === 0 || output.includes('.bg-red')) {
        expect(output).toContain('.bg-red')
        break
      }
    }
  })

  it('supports unocss.config.js cli options', async () => {
    const testDir = getTestDir()
    const outFiles = ['./uno1.css', './test/uno2.css']
    const files = [
      {
        path: 'views/index1.html',
        content: '<div class="bg-blue"></div>',
      },
      {
        path: 'views/index2.html',
        content: '<div class="bg-red"></div>',
      },
      {
        path: 'unocss.config.js',
        content: `
import { defineConfig } from 'unocss'
export default defineConfig({
    cli: {
    entry: [
      {
        patterns: ['views/index1.html'],
        outFile: '${outFiles[0]}',
      },
      {
        patterns: ['views/index2.html'],
        outFile: '${outFiles[1]}',
      },
    ],
  }
})
          `.trim(),
      },
    ]
    await Promise.all(files.map(({ path, content }) => fs.outputFile(resolve(testDir, path), content)))
    await runAsyncChildProcess(testDir, '', '')

    await sleep(500)
    const [output1, output2] = await Promise.all(outFiles.map(async file => readFile(testDir, resolve(testDir, file))))
    expect(output1).toContain('.bg-blue')
    expect(output2).toContain('.bg-red')
  })
})
// ----- Utils -----
function sleep(time = 300) {
  return new Promise<void>((resolve) => {
    setTimeout(() => {
      resolve()
    }, time)
  })
}

function getTestDir() {
  return resolve(tempDir, Math.round(Math.random() * 100000).toString())
}

function initOutputFiles(testDir: string, files: Record<string, string>) {
  return Promise.all(
    Object.entries(files).map(([path, content]) =>
      fs.outputFile(resolve(testDir, path), content, 'utf8'),
    ),
  )
}

function runAsyncChildProcess(cwd: string, ...args: string[]) {
  return startCli(cwd, ['', '', ...args, '--no-preflights'])
}

function readFile(testDir: string, targetFile?: string) {
  return fs.readFile(resolve(testDir, targetFile ?? 'uno.css'), 'utf8')
}

async function runCli(files: Record<string, string>, transformFile?: string) {
  const testDir = getTestDir()

  await initOutputFiles(testDir, files)
  await runAsyncChildProcess(testDir, 'views/**/*')

  const output = await readFile(testDir)

  if (transformFile) {
    const transform = await readFile(testDir, transformFile)
    return {
      output,
      transform,
    }
  }

  return {
    output,
  }
}
