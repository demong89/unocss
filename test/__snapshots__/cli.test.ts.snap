// Vitest Snapshot v1

exports[`cli > builds uno.css 1`] = `
"/* layer: default */
.max-w-screen-md{max-width:768px;}
.p-4{padding:1rem;}"
`;

exports[`cli > supports directives transformer 1`] = `""`;

exports[`cli > supports directives transformer 2`] = `".btn-center{margin-top:0;margin-bottom:0;text-align:center;font-weight:500;}"`;

exports[`cli > supports unocss.config.js 1`] = `
"/* layer: shortcuts */
.box{margin-left:auto;margin-right:auto;max-width:80rem;border-radius:0.375rem;--un-bg-opacity:1;background-color:rgba(243,244,246,var(--un-bg-opacity));padding:1rem;--un-shadow:var(--un-shadow-inset) 0 1px 2px 0 var(--un-shadow-color, rgba(0,0,0,0.05));box-shadow:var(--un-ring-offset-shadow), var(--un-ring-shadow), var(--un-shadow);}"
`;

exports[`cli > supports variantGroup transformer 1`] = `
"/* layer: default */
.border-red{--un-border-opacity:1;border-color:rgba(248,113,113,var(--un-border-opacity));}
.border-solid{border-style:solid;}
.p-4{padding:1rem;}"
`;

exports[`cli > supports variantGroup transformer 2`] = `"<div class=\\"p-4 border-solid border-red\\"></div>"`;
